### PR TITLE
[Raft] Implement RequestVote

### PR DESCRIPTION
**Pull request by liana-yang**

### CODE REVIEW

File: lab/src/raft/request_vote.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
+package raft
```
Comment (flyingmeat) [inline lab/src/raft/request_vote.go:1]: Split the file Hoye!

File: lab/src/raft/request_vote.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	if args.Term < rf.currentTerm {
+		reply.VoteGranted = false
+		return
+	}
+
+	// If votedFor is null or candidateId, and candidate’s log is at least as up-to-date as receiver’s log, grant vote (§5.2, §5.4)
+	if rf.votedFor == -1 || rf.votedFor == args.CandidateId {
+		localLastLog := getLastLog(rf.log)
+		isLogUpToDate := localLastLog.Term < args.LastLogTerm
+		if localLastLog.Term == args.LastLogTerm {
```
Comment (flyingmeat) [inline lab/src/raft/request_vote.go]: You can combine this line with previous line

File: lab/src/raft/request_vote.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	//  Reply false if term < currentTerm (§5.1)
+	if args.Term < rf.currentTerm {
+		reply.VoteGranted = false
+		return
+	}
+
+	// If votedFor is null or candidateId, and candidate’s log is at least as up-to-date as receiver’s log, grant vote (§5.2, §5.4)
+	if rf.votedFor == -1 || rf.votedFor == args.CandidateId {
+		localLastLog := getLastLog(rf.log)
+		isLogUpToDate := localLastLog.Term < args.LastLogTerm
```
Comment (flyingmeat) [inline lab/src/raft/request_vote.go]: What if localLastLog is nil here. I think we might need to check this before the comparing

Comment (liana-yang) [reply lab/src/raft/request_vote.go]: Good point. The log should not be up-to-date if it's nil.

File: lab/src/raft/request_vote.go (hunk 2 above)
Comment (liana-yang) [reply lab/src/raft/request_vote.go]: I am pretty indifferent, but I think it's pretty clear that only will the index of the logs be compared when their term are equal.

File: lab/src/raft/request_vote.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+	//  Reply false if term < currentTerm (§5.1)
+	if args.Term < rf.currentTerm {
+		return
+	}
+
+	// If votedFor is null or candidateId, and candidate’s log is at least as up-to-date as receiver’s log, grant vote (§5.2, §5.4)
+	if rf.votedFor == -1 || rf.votedFor == args.CandidateId {
+		localLastLog := getLastLog(rf.log)
+		if localLastLog == nil {
```
Comment (flyingmeat) [inline lab/src/raft/request_vote.go]: if localLastLog is nil, I think then any other candidate's log should be more up-to-date than requested server

Comment (liana-yang) [reply lab/src/raft/request_vote.go]: Done.

File: lab/src/raft/request_vote.go (hunk 2 above)
Comment (flyingmeat) [reply lab/src/raft/request_vote.go]: Hmm. Yeah, my bad, this is actually very clear!

File: lab/src/raft/request_vote.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	//  Reply false if term < currentTerm (§5.1)
+	if args.Term < rf.currentTerm {
+		return
+	}
+
+	// If votedFor is null or candidateId, and candidate’s log is at least as up-to-date as receiver’s log, grant vote (§5.2, §5.4)
+	if rf.votedFor == -1 || rf.votedFor == args.CandidateId {
+		localLastLog := getLastLog(rf.log)
+		isLogUpToDate := localLastLog == nil || localLastLog.Term < args.LastLogTerm
+		if localLastLog.Term == args.LastLogTerm {
```
Comment (flyingmeat) [inline lab/src/raft/request_vote.go]: `if localLastLog != nil && localLastLog.Term == args.LastLogTerm`

Comment (liana-yang) [reply lab/src/raft/request_vote.go]: Good catch!

**Review by flyingmeat — APPROVED** [review]

**Pull request merged**